library/src/G4CMPPhononBoundaryProcess.cc

```diff
@@ -192,128 +227,249 @@ DoReflection(const G4Track& aTrack, const G4Step& aStep,
-		JustWarning, ("Phonon "+refltype+" reflection failed").c_str());
-    DoSimpleKill(aTrack, aStep, aParticleChange);
+		JustWarning, ("Phonon "+refltype+" reflection failed"+"\nPhonon mode at time of death: "+G4PhononPolarization::Label(mode)).c_str());
+    DoSimpleKill(aTrack, aStep, particleChange);
     return;
   }
-
-  // SANITY CHECK:  Project a 1 um step in the new direction, see if it
-  // is still in the correct (pre-step) volume.
-
-  if (verboseLevel>2) {
-    G4ThreeVector stepPos = surfacePoint + 1*um * vdir;
-
-    G4cout << " New travel direction " << vdir
-	   << "\n from " << surfacePoint << "\n   to " << stepPos << G4endl;
-
-    G4ThreeVector stepLocal = GetLocalPosition(stepPos);
-    G4VSolid* solid = aStep.GetPreStepPoint()->GetPhysicalVolume()->GetLogicalVolume()->GetSolid();
-
-    EInside place = solid->Inside(stepLocal);
-    G4cout << " After trial step, " << (place==kInside ? "inside"
-					: place==kOutside ? "OUTSIDE"
-					: "on surface") << G4endl;
-  }
-
-  trackInfo->SetWaveVector(reflectedKDir);
-  particleChange.ProposeVelocity(v);
-  particleChange.ProposeMomentumDirection(vdir);
 }
 
-
 // Generate specular reflection corrected for momentum dispersion
 
 G4ThreeVector G4CMPPhononBoundaryProcess::
-GetReflectedVector(const G4ThreeVector& waveVector,
-		   const G4ThreeVector& surfNorm, G4int mode) const {
-  // Specular reflecton should reverses momentum along normal
+GetSpecularVector(const G4ThreeVector& waveVector,
+                  G4ThreeVector& surfNorm, G4int mode,
+                  G4ThreeVector& surfacePoint) {
+  // Specular reflecton should reverse momentum along normal
   G4ThreeVector reflectedKDir = waveVector.unit();
-  G4double kPerp = reflectedKDir * surfNorm;
-  (reflectedKDir -= 2.*kPerp*surfNorm).setMag(1.);
+  G4double kPerp = reflectedKDir * surfNorm;		// Dot product between k and norm
+  (reflectedKDir -= 2.*kPerp*surfNorm).setMag(1.);	// Reflect against normal
+
+  if (G4CMP::PhononVelocityIsInward(theLattice,mode,reflectedKDir,surfNorm,
+                                    surfacePoint))
+    return reflectedKDir;
+
+  // Reflection didn't work as expected, need to correct:
+  // If the reflected wave vector cannot propagate in the bulk
+  // (i.e., the reflected k⃗ has an associated v⃗g which is not inwardly directed.)
+  // That surface wave will propagate until it reaches a point
+  // where the wave vector has an inwardly directed v⃗g.
+  RotateToLocalDirection(reflectedKDir);
+  G4ThreeVector newNorm = surfNorm;
+  RotateToLocalDirection(newNorm);
   
-  if (verboseLevel>2) {
-    G4cout << " specular reflection with normal " << surfNorm
-	   << "\n Perpendicular wavevector " << kPerp*surfNorm
-	   << " (mag " << kPerp << ")" << G4endl;
+  // Initialize solid object and utilities
+  G4VSolid* solid = GetCurrentVolume()->GetLogicalVolume()->GetSolid();
+  G4CMPSolidUtils* solidUtils = new G4CMPSolidUtils(solid, verboseLevel,
```

Comment:
This does not need to be a pointer.  If you make it just a local variable, i.e.,
```
G4CMPSolidUtils solidUtils(solid, verboseLevel, GetProcessName());
```
then you don't have to `delete` it later.  It'll just go away when the function returns.